src/bvh/bvh_node.rs

```diff
@@ -0,0 +1,608 @@
+    ) {
+        if let Some((left, right)) = Self::prep_build(args) {
+            // Self::build(left);
+            // Self::build(right);
+            executor(left, right);
+            // RayonBuildStrategy::dispatch(|| Self::build(left), || Self::build(right), left.indices.len() + right.indices.len(), left.depth as usize)
+        }
+    }
+
+    /// Builds a [`BvhNode`] recursively using SAH partitioning.
+    ///
+    /// [`BvhNode`]: enum.BvhNode.html
+    ///
+    pub fn prep_build<S: BHShape<T, D>>(
+        args: BvhNodeBuildArgs<S, T, D>,
+    ) -> Option<(BvhNodeBuildArgs<S, T, D>, BvhNodeBuildArgs<S, T, D>)> {
+        let BvhNodeBuildArgs {
+            shapes,
+            indices,
+            nodes,
+            parent_index,
+            depth,
+            node_index,
+            aabb_bounds,
+            centroid_bounds,
+        } = args;
+        // If there is only one element left, don't split anymore
+        if indices.len() == 1 {
+            let shape_index = indices[0];
+            nodes[0].write(BvhNode::Leaf {
+                parent_index,
+                shape_index: shape_index.0,
+            });
+            // Let the shape know the index of the node that represents it.
+            shapes.set_node_index(shape_index, node_index);
+            return None;
+        }
+
+        // Find the axis along which the shapes are spread the most.
+        let split_axis = centroid_bounds.largest_axis();
+        let split_axis_size = centroid_bounds.max[split_axis] - centroid_bounds.min[split_axis];
+
+        // The following `if` partitions `indices` for recursively calling `Bvh::build`.
+        let (
+            (child_l_aabb, child_l_centroid, child_l_indices),
+            (child_r_aabb, child_r_centroid, child_r_indices),
+        ) = if split_axis_size < T::epsilon() {
+            // In this branch the shapes lie too close together so that splitting them in a
+            // sensible way is not possible. Instead we just split the list of shapes in half.
+            let (child_l_indices, child_r_indices) = indices.split_at_mut(indices.len() / 2);
+            let (child_l_aabb, child_l_centroid) = joint_aabb_of_shapes(child_l_indices, shapes);
+            let (child_r_aabb, child_r_centroid) = joint_aabb_of_shapes(child_r_indices, shapes);
+
+            (
+                (child_l_aabb, child_l_centroid, child_l_indices),
+                (child_r_aabb, child_r_centroid, child_r_indices),
+            )
+        } else {
+            BvhNode::build_buckets(
+                shapes,
+                indices,
+                split_axis,
+                split_axis_size,
+                &centroid_bounds,
+                &aabb_bounds,
+            )
+        };
+
+        let left_len = child_l_indices.len() * 2 - 1;
+        let child_l_index = node_index + 1;
+        let child_r_index = child_l_index + left_len;
```

Comment:
Document this: What are you doing here and what are the numbers?

Reply:
Done